backends/graphics/opengl/pipelines/libretro.cpp

```diff
@@ -115,17 +115,84 @@ static const char *g_compatFragment =
 		"#endif\n"
 	"#endif\n";
 
+/* This TextureTarget handles the scaling of coordinates from the window coordinates space
+ * to the input coordinates space */
+class LibRetroTextureTarget : public TextureTarget {
+public:
+	bool setScaledSize(uint width, uint height, const Common::Rect &scalingRect) {
+		GLTexture *texture = getTexture();
+		if (!texture->setSize(width, height)) {
+			return false;
+		}
+
+		const uint texWidth  = texture->getWidth();
+		const uint texHeight = texture->getHeight();
+
+		// Set viewport dimensions.
+		_viewport[0] = 0;
+		_viewport[1] = 0;
+		_viewport[2] = texWidth;
+		_viewport[3] = texHeight;
+
+		const float ratioW = (float)width  / scalingRect.width();
+		const float ratioH = (float)height / scalingRect.height();
+
+		// Setup scaling projection matrix.
+		// This projection takes window screen coordinates and converts it to input coordinates normalized
+		_projectionMatrix(0, 0) = 2.f * ratioW / texWidth;
+		_projectionMatrix(0, 1) = 0.f;
+		_projectionMatrix(0, 2) = 0.f;
+		_projectionMatrix(0, 3) = 0.f;
+
+		_projectionMatrix(1, 0) = 0.f;
+		_projectionMatrix(1, 1) = 2.f * ratioH / texHeight;
+		_projectionMatrix(1, 2) = 0.f;
+		_projectionMatrix(1, 3) = 0.f;
+
+		_projectionMatrix(2, 0) = 0.f;
+		_projectionMatrix(2, 1) = 0.f;
+		_projectionMatrix(2, 2) = 0.f;
+		_projectionMatrix(2, 3) = 0.f;
+
+		_projectionMatrix(3, 0) = -1.f - (2.f * scalingRect.left) * ratioW / texWidth;
+		_projectionMatrix(3, 1) = -1.f - (2.f * scalingRect.top) * ratioH / texHeight;
+		_projectionMatrix(3, 2) = 0.0f;
+		_projectionMatrix(3, 3) = 1.0f;
+
+		// Directly apply changes when we are active.
+		if (isActive()) {
+			applyViewport();
+			applyProjectionMatrix();
+		}
+		return true;
+	}
+};
+
+/* This Pipeline is used by the Framebuffer objects in the LibRetro passes.
+ * It does nothing as everything is already handled by us. */
+class FakePipeline : public Pipeline {
+public:
+	FakePipeline() {}
+	void setColor(GLfloat r, GLfloat g, GLfloat b, GLfloat a) override { }
+	void drawTexture(const GLTexture &texture, const GLfloat *coordinates, const GLfloat *texcoords) override { }
+	void setProjectionMatrix(const Math::Matrix4 &projectionMatrix) override {}
+protected:
+	void activateInternal() override {}
```

Comment:
Is it also necessary to override `deactivateInternal()` as well?

Reply:
Ah yes, it would be cleaner. Thanks!